src/shapes.ts

```diff
@@ -0,0 +1,270 @@
+/**
+ * @packageDocumentation
+ * @module settings
+ *
+ * This file consists of shapes supported by chemiscope via 3dmol
+ *
+ * To add a new shape type, extend the `Shape` class and define the
+ * inherited functions (`validateParams` and `outputTo3Dmol`).
+ * `validateParams` should check a `ShapeData` object to see that
+ * it has the correct parameters for the given type. `outputTo3Dmol`
+ * should take a color and output a `CustomShapeSpec` (https://3dmol.csb.pitt.edu/doc/types.html#CustomShapeSpec)
+ * object. You will also need to add any parameters for your shape
+ * to ShapeData.
+ *
+ * To add a new shape, you will also need to add code to `structure/viewer.ts`.
+ */
+
+import { default as $3Dmol } from '3dmol';
+import { Quaternion } from '3dmol';
+
+// Just an interface to enforce XYZ-type coordinates
+export interface XYZ {
+    x: number;
+    y: number;
+    z: number;
+}
+
+// A base class for storing shape data in ase files
+export interface ShapeData {
+    radius?: number;
+    semiaxes?: [number, number, number];
+    orientation?: [number, number, number, number];
+    kind: string;
+}
+
+export class Shape {
+    /// orientation of the particle
+    public position: [number, number, number];
+    public quaternion: Quaternion;
+
+    constructor(pos: [number, number, number], orientation: [number, number, number, number]) {
+        this.quaternion = new Quaternion(
+            orientation[1],
+            orientation[2],
+            orientation[3],
+            orientation[0]
+        );
+        this.position = pos;
+    }
+
+    // disabling eslint because it complains parameters is never used
+    // function to check the parameters corresponding to the given
+    // sub-class exist.
+    /* eslint-disable-next-line */
+    public static validateParams(parameters: ShapeData): boolean {
+        return true;
+    }
+    // function to output a `CustomShapeSpec` for the given object
+    public outputTo3Dmol(color: $3Dmol.ColorSpec): $3Dmol.CustomShapeSpec {
+        return { color: color };
+    }
+}
+
+export class Sphere extends Shape {
+    public radius: number;
+
+    constructor(radius: number, pos: [number, number, number] = [0, 0, 0]) {
+        super(pos, [1, 0, 0, 0]);
+        this.radius = radius;
+    }
+
+    public static validateParams(parameters: ShapeData): boolean {
+        if ('radius' in parameters) {
+            const radius = parameters['radius'];
+            if (typeof radius === 'number') {
+                return true;
+            } else {
+                return false;
+            }
+        } else {
+            return false;
+        }
+    }
+    /**
+     ** Following https://computergraphics.stackexchange.com/questions/7426/triangulation-of-vertices-of-an-ellipsoid
+     **/
+    public outputTo3Dmol(color: $3Dmol.ColorSpec, resolution: number = 20): $3Dmol.CustomShapeSpec {
```

Comment:
Why use a custom triangulation for spheres as well?

Reply:
Oh this can go! I just left it in for example-sake.